### PR TITLE
Implement Draft Pull Request Feature #2319

### PR DESCRIPTION
**Pull request by joobisb**
### Before submitting a pull-request to GitBucket I have first:

- [x] read the [contribution guidelines](https://github.com/gitbucket/gitbucket/blob/master/.github/CONTRIBUTING.md)
- [x] rebased my branch over master
- [x] verified that project is compiling
- [x] verified that tests are passing
- [x] squashed my commits as appropriate *(keep several commits if it is relevant to understand the PR)*
- [x] [marked as closed using commit message](https://help.github.com/articles/closing-issues-via-commit-messages/) all issue ID that this PR should correct

This PR is related to issue #2319 

This PR enables users to either create a Pull request or a Draft Request. If a draft request is being created, it cannot be merged until it as marked as ```Ready for Review```. So a WIP code can be raised as a PR, so that early review is possible for peer developers.

image

image

image

### CODE REVIEW

File: src/main/twirl/gitbucket/core/pulls/mergeguide.scala.html [hunk 1]
```diff
@@ -199,6 +222,15 @@
...
 
 <script>
+
+$(function(){
+  $('#ready-for-review').click(function(){
+    $.post('@helpers.url(originRepository)/pull/@issue.issueId/update_draft', function(data, status){
+      location.reload();
+    })
+  });
+});
```
Comment (takezoe) [inline src/main/twirl/gitbucket/core/pulls/mergeguide.scala.html]: If you want to submit a request by form, you don't need to send a request by Ajax. In opposite, if you want to send a request by Ajax, you don't need to define `<form>` and form submission related attributes like `validate` and `formaction` in `<button>`.

File: src/main/twirl/gitbucket/core/pulls/mergeguide.scala.html [hunk 2]
```diff
@@ -77,23 +77,45 @@
...
+                 </div>
+               } else {
+                   <form method="POST" action="@context.path/@originRepository.owner/@originRepository.name/pull/@issue.issueId/update_draft" validate="true">
+                   <span class="strong">This pull request is still a work in progress.</span>
+                   <div class="small">
+                   Draft pull requests cannot be merged.
+                   </div>
+                   <div class="pull-right">
+                     <input type="button" class="btn btn-default" value="Ready for review" id="ready-for-review" validate="true" formaction="@context.path/@originRepository.owner/@originRepository.name/pull/@issue.issueId/update_draft"/>
+                   </div>`
```
Comment (takezoe) [inline src/main/twirl/gitbucket/core/pulls/mergeguide.scala.html]: Could you remove ending `?

File: src/main/twirl/gitbucket/core/pulls/mergeguide.scala.html [hunk 3]
```diff
@@ -77,23 +77,45 @@
...
-              <span class="strong">Merging can be performed automatically.</span>
-              <div class="small">
-                Merging can be performed automatically.
-              </div>
+              @if(!pullreq.isDraft){
+                  <span class="strong">Merging can be performed automatically.</span>
+                  <div class="small">
+                  Merging can be performed automatically.
+                 </div>
+               } else {
```
Comment (takezoe) [inline src/main/twirl/gitbucket/core/pulls/mergeguide.scala.html]: This part looks almost the same as below part. I think this code duplication can be avoided by devising conditional branching.

File: src/main/twirl/gitbucket/core/pulls/compare.scala.html [hunk 4]
```diff
@@ -159,6 +172,21 @@ <h4>There isn't anything to compare.</h4>
   }
 }
 <script>
+
+$(function(){
+  $('#draft-request').click(function(){
+    $("#is-draft").val(true);
+    $('#submit-button').attr('value', 'Create draft request')
+    $('#action').val('create draft request');
```
Comment (takezoe) [inline src/main/twirl/gitbucket/core/pulls/compare.scala.html]: Is this necessary?

File: src/main/twirl/gitbucket/core/pulls/compare.scala.html [hunk 5]
```diff
@@ -159,6 +172,21 @@ <h4>There isn't anything to compare.</h4>
...
+  });
+  $('#pull-request').click(function(){
+  $("#is-draft").val(false);
+    $('#submit-button').attr('value', 'Create pull request')
+    $('#action').val('create pull request');
```
Comment (takezoe) [inline src/main/twirl/gitbucket/core/pulls/compare.scala.html]: Is this necessary?

File: src/main/scala/gitbucket/core/controller/PullRequestsController.scala [hunk 6]
```diff
@@ -338,14 +342,26 @@ trait PullRequestsControllerBase extends ControllerBase {
     }) getOrElse NotFound()
   })
 
+  post("/:owner/:repository/pull/:id/update_draft")(readableUsersOnly { baseRepository =>
+    (for {
+      issueId <- params("id").toIntOpt
+      (_, pullreq) <- getPullRequest(baseRepository.owner, baseRepository.name, issueId)
+      owner = pullreq.requestUserName
+      name = pullreq.requestRepositoryName
+      if hasDeveloperRole(owner, name, context.loginAccount)
```
Comment (takezoe) [inline src/main/scala/gitbucket/core/controller/PullRequestsController.scala]: Adding constraint `hasDeveloperRole` means that only users who has write permission to the repository can release the draft status. This also means the author of pull request may not be able to do that. Is this ok?

**Comment by joobisb** [comment]
> @joobisb Thanks for the great contribution! Could you check my comments?

@takezoe sure, will look at it

File: src/main/twirl/gitbucket/core/pulls/mergeguide.scala.html (hunk 1 above)
Comment (joobisb) [reply src/main/twirl/gitbucket/core/pulls/mergeguide.scala.html]: ok, will change acccordingly..

File: src/main/twirl/gitbucket/core/pulls/mergeguide.scala.html (hunk 2 above)
Comment (joobisb) [reply src/main/twirl/gitbucket/core/pulls/mergeguide.scala.html]: ok

File: src/main/twirl/gitbucket/core/pulls/mergeguide.scala.html (hunk 3 above)
Comment (joobisb) [reply src/main/twirl/gitbucket/core/pulls/mergeguide.scala.html]: ok, will avoid the duplication

File: src/main/twirl/gitbucket/core/pulls/compare.scala.html (hunk 4 above)
Comment (joobisb) [reply src/main/twirl/gitbucket/core/pulls/compare.scala.html]: no, will remove it.

File: src/main/twirl/gitbucket/core/pulls/compare.scala.html (hunk 5 above)
Comment (joobisb) [reply src/main/twirl/gitbucket/core/pulls/compare.scala.html]: no, will remove

File: src/main/scala/gitbucket/core/controller/PullRequestsController.scala (hunk 6 above)
Comment (joobisb) [reply src/main/scala/gitbucket/core/controller/PullRequestsController.scala]: @takezoe Here the owner is the base repository owner, so the author of pull request can release the draft status. Also users having write permission can also release the draft status.

**Comment by joobisb** [comment]
> @joobisb Thanks for the great contribution! Could you check my comments?

@takezoe I've addressed your comments. Please have a look.

**Pull request merged**